STYLE.md

```diff
@@ -0,0 +1,262 @@
+# Vector Style Guide
+
+As a large, open-source project, it can be a struggle to ensure a consistent level of code quality.  This style guide is
+meant to be the canonical reference for all things style: code comments, formatting, acceptable (or unacceptable)
+crates, data structures, or algorithms, and so on.
+
+In essence, we hope to turn pull request review comments like "why did you do it this way?" or "I think you could try
+doing it this way" into "we always do X this way: <link to style guide>".
+
+# Formatting
+
+At a high-level, code formatting is straightforward: we use the native `rustfmt` exclusively.
+
+You can acquire `rustfmt` -- which is invoked as `cargo fmt` -- by following the directions listed out [in the rustfmt
+repository](https://github.com/rust-lang/rustfmt#on-the-stable-toolchain).
+
+Vector has its own formatting rules (`.rustfmt.toml`) that will automatically be used when you run `cargo fmt` within
+the repository.  If you're ever in doubt, you can also run `make check-fmt` which will invoke `cargo fmt` and ensure all
+code follows our formatting rules.
+
+# Code Organization
+
+Code is primarily split into two main directories: `lib/` and `src/`.
+
+## `lib/`: shared libraries, etc
+
+We use `lib` almost entirely for shared libraries and for isolating specific pieces of code. As Vector itself involves a
+large number of dependencies, it can be beneficial to move code into isolated crates under `lib/` in order to allow them
+not only to be shared, but also to reduce the amount of code that must be processed by helper tools like `cargo check`
+and `rust-analyzer` during normal development, which in turn speeds up the feedback loop between writing code and
+getting informed about errors, warnings, and so on.
+
+## `src/`: main binary and all related functionality
+
+The bulk of code resides in the `src/` directory/crate, and within that, is almost entirely related to the
```

Comment:
Yeah, I guess the thrust of the statement is meant to refer to the user-facing code -- components, configuration parsing, and so on. I'll think about how to better reword this to be along the lines of a code organization _map_: rather than how much code lives where, more of a... "if you're looking for X, you should look here", etc.